### PR TITLE
add auto code formatting hook

### PR DESCRIPTION
**Pull request by jthistle**
This needs to be installed with a script in the new `hooks` directory.

Please also note that a line in the config which is not available in the version of Uncrustify in the Ubuntu package repo has been commented out. That shouldn't be a problem, because it was the default value anyway.

### CODE REVIEW

File: hooks/install.sh [hunk 1]
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env bash
+
+HOOKSDIR=./.git/hooks/
+
+cp ./hooks/pre-commit "$HOOKSDIR/pre-commit"
```
Comment (njvdberg) [inline hooks/install.sh:5]: Not better of add that newline at the end?

File: hooks/uninstall.sh [hunk 2]
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env bash
+
+HOOKSDIR=./.git/hooks/
+
+rm "$HOOKSDIR/pre-commit"
```
Comment (njvdberg) [inline hooks/uninstall.sh:5]: Same, newline missing.

File: hooks/pre-commit [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+echo "Uncrustifying staged files..."
+
+RELPATH=../..
+staged=$(git diff --name-only --cached)
+
+cd ./tools/codestyle/
+while IFS= read -r file; do
+	ext="${file##*.}"
+	if [ $ext != "cpp" ] && [ $ext != "h" ]; then
```
Comment (Jojo-Schmitz) [inline hooks/pre-commit:11]: Wouldn't we use the 4 space indenting here too?

File: tools/codestyle/uncrustify_musescore.cfg [hunk 4]
```diff
@@ -248,7 +248,7 @@ sp_assign                                = ignore   # ignore/add/remove/force
 sp_cpp_lambda_assign                     = remove   # ignore/add/remove/force
 
 # Add or remove space after the capture specification in C++11 lambda.
-sp_cpp_lambda_square_paren               = ignore   # ignore/add/remove/force
+# sp_cpp_lambda_square_paren               = ignore   # ignore/add/remove/force
```
Comment (Jojo-Schmitz) [inline tools/codestyle/uncrustify_musescore.cfg:251]: was that an intended change? Seems unrelated to the issue this PR tries to solve

File: hooks/install.sh (hunk 1 above)
Comment (jthistle) [reply hooks/install.sh:5]: having a newline at the end of a file is of practically no importance

File: hooks/pre-commit (hunk 3 above)
Comment (jthistle) [reply hooks/pre-commit:11]: for shell scripts, I've always used tabs. I thought our coding rules apply only to c/cpp :man_shrugging:

File: tools/codestyle/uncrustify_musescore.cfg (hunk 4 above)
Comment (jthistle) [reply tools/codestyle/uncrustify_musescore.cfg:251]: See the note in the PR body: 

>Please also note that a line in the config which is not available in the version of Uncrustify in the Ubuntu package repo has been commented out. That shouldn't be a problem, because it was the default value anyway.

File: hooks/install.sh (hunk 1 above)
Comment (Jojo-Schmitz) [reply hooks/install.sh:5]: Is looks ugly, enough reason IMHO

File: tools/codestyle/uncrustify_musescore.cfg (hunk 4 above)
Comment (Jojo-Schmitz) [reply tools/codestyle/uncrustify_musescore.cfg:251]: For now at least. Other she'll scripts don't use tabs though

File: hooks/install.sh (hunk 1 above)
Comment (njvdberg) [reply hooks/install.sh:5]: Back in the old days, I used to have some problems with the missing last newline ;-)

File: hooks/pre-commit (hunk 3 above)
Comment (Jojo-Schmitz) [reply hooks/pre-commit:11]: For now at least. Other she'll scripts don't use tabs though

File: tools/codestyle/uncrustify_musescore.cfg (hunk 4 above)
Comment (Jojo-Schmitz) [reply tools/codestyle/uncrustify_musescore.cfg:251]: Ah, missed that

File: hooks/install.sh (hunk 1 above)
Comment (jthistle) [reply hooks/install.sh:5]: >It looks ugly

lol, only in the way Github displays it ;)

File: hooks/pre-commit (hunk 3 above)
Comment (jthistle) [reply hooks/pre-commit:11]: `build/run_tests.sh` uses 2 spaces. `build/Linux+BSD/mscore.1.preview.sh` uses a mix of tabs and 4 spaces. I think it's safe to say that we don't have a code style for shell scripts.

Comment (Jojo-Schmitz) [reply hooks/pre-commit:11]: Then it seems about time to establish one. As we use 4 spaces indent in code, we should in scripts too

Comment (jthistle) [reply hooks/pre-commit:11]: That's a PR for you to make, then :smiley:
